sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -170,34 +170,31 @@ namespace Azure { namespace Core { namespace Http {
     int m_port{-1};
     std::string m_encodedPath;
     std::string m_fragment;
-    // query parameters are all decoded
-    std::map<std::string, std::string> m_queryParameters;
-    std::map<std::string, std::string> m_retryQueryParameters;
+    // query parameters are all encoded
+    std::map<std::string, std::string> m_encodedQueryParameters;
+    std::map<std::string, std::string> m_retryEncodedQueryParameters;
     bool m_retryModeEnabled{false};
 
-    /*********  private static methods for all instances *******/
-    static std::string Decode(const std::string& value);
-    static std::string EncodeHost(const std::string& host);
-    static std::string EncodePath(const std::string& path);
-    static std::string EncodeQuery(const std::string& query);
-    static std::string EncodeFragment(const std::string& fragment);
-    static std::string EncodeImpl(
-        const std::string& source,
-        const std::function<bool(int)>& should_encode);
+    // List of default non-url-encode chars. While url-encoding string, do not scape any char on
+    // this set
+    static std::unordered_set<unsigned char> defaultNonUrlEncodeChars;
 
     void StartRetry()
     {
       m_retryModeEnabled = true;
-      m_retryQueryParameters.clear();
+      m_retryEncodedQueryParameters.clear();
     }
 
     void StopRetry()
     {
       m_retryModeEnabled = false;
-      m_retryQueryParameters.clear();
+      m_retryEncodedQueryParameters.clear();
     }
 
   public:
+    static std::string Decode(const std::string& value);
+    static std::string Encode(const std::string& value, const char* extraSymbolsToEncode = "");
```

Comment:
This API is really not friendly to use.

Reply:
It reduces the concept count, with the goal to have the default be the most commonly use, and avoid having overloads specific for encoding various components.

The caller only needs to provide the extra characters that they might need to encode, beyond the standard unreserved list.

This is fairly similar to the .NET Url encoding method:
https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UrlEncoder.cs#L59-L66

What would you recommend instead?

Reply:
So in storage code, we should call `Url::AppendQuery(key, Encode(value, "="));` because we want `=` to be encoded.
okay, I see what you mean. I think it can work.

Reply:
My proposal is adding two more functions,

`EncodePath` which will call `Encode` without any extra symbols to encode

`EncodeQueryParameterValue` which will call `Encode` with `=` as extra symbol

Reply:
@Jinming-Hu , I don't think we really need those 2 extra functions.  `=` is encoded by default.

Reply:
fine